app/jobs/inboxes/fetch_imap_emails_job.rb

```diff
@@ -41,9 +46,51 @@ def fetch_mail_for_channel(channel)
     end
   end
 
+  def fetch_mail_for_ms_provider(channel)
+    return unless channel.microsoft? && channel.provider_config[:access_token].blank?
+
+    access_token = valid_access_token channel
+
+    return unless access_token
+
+    imap = imap_authenticate(channel, access_token)
+
+    process_mails(imap, channel)
+  end
+
+  def process_mails(imap, channel)
+    imap.search(['BEFORE', tomorrow, 'SINCE', yesterday]).each do |message_id|
+      inbound_mail = Mail.read_from_string imap.fetch(message_id, 'RFC822')[0].attr['RFC822']
+
+      next if channel.inbox.messages.find_by(source_id: inbound_mail.message_id).present?
+
+      process_mail(inbound_mail, channel)
+    end
+  end
+
+  def imap_authenticate(channel, access_token)
+    imap = Net::IMAP.new(channel.imap_address, channel.imap_port, true)
+    imap.authenticate('XOAUTH2', channel.imap_login, access_token)
+    imap.select('INBOX')
+    imap
+  end
+
+  def yesterday
+    (Time.zone.today - 1).strftime('%d-%b-%Y')
+  end
+
+  def tomorrow
+    (Time.zone.today + 1).strftime('%d-%b-%Y')
+  end
+
   def process_mail(inbound_mail, channel)
     Imap::ImapMailbox.new.process(inbound_mail, channel)
   rescue StandardError => e
     ChatwootExceptionTracker.new(e, account: channel.account).capture_exception
   end
+
+  # Making sure the access token is valid for microsoft provider
+  def valid_access_token(channel)
```

Comment:
We will be doing this step every time when we fetch the mail right ? ( in 1 minute ). In case we won't be needing to run the job separately in a schedule ?

Reply:
Oh yes! you are right :D.. This is happening every 5 minute to check if the token is still valid and if not it will request for new token.

Removing the refresh token job.